dvc/data/add.py

```diff
@@ -0,0 +1,58 @@
+from typing import TYPE_CHECKING, Tuple, Union
+
+from .stage import stage
+from .transfer import transfer
+from .tree import Tree
+
+if TYPE_CHECKING:
+    from dvc.data.meta import Meta
+    from dvc.fs.base import FileSystem
+    from dvc.ignore import DvcIgnore
+    from dvc.objects.db import ObjectDB
+    from dvc.objects.file import HashFile
+    from dvc.types import AnyPath
+
+
+def add(
+    odb: "ObjectDB",
+    fs: "FileSystem",
+    fs_path: "AnyPath",
+    name: str,
+    filter_prefix: Union["AnyPath", Tuple[str, ...]] = None,
+    hardlink: bool = False,
+    dvcignore: "DvcIgnore" = None,
+    upload: bool = False,
+    jobs: int = None,
+    no_progress_bar: bool = False,
+) -> Tuple["Meta", Union["HashFile", "Tree"]]:
+    staging, meta, obj = stage(
+        odb,
+        fs_path,
+        fs,
+        name,
+        upload=upload,
+        jobs=jobs,
+        dvcignore=dvcignore,
+        no_progress_bar=no_progress_bar,
+    )
+
+    obj_ids = {obj.hash_info}
+
+    if filter_prefix is None or isinstance(filter_prefix, tuple):
+        prefix = filter_prefix
+    else:
+        prefix = fs.path.parts(fs.path.relpath(filter_prefix, fs_path))
+
+    if prefix and isinstance(obj, Tree):
+        obj = obj.filter(prefix)
+        obj_ids.update({oid for _, _, oid in obj})
+
+    transfer(
+        staging,
+        odb,
+        obj_ids,
+        shallow=bool(prefix),
```

Comment:
I am still trying to understand `shallow` here. Let me know if the `shallow` should be a separate option or only set on `prefix` as done here.

Reply:
My understanding is that it ignores the content of the `Tree`'s _leaf_ object. So, if we pass the content of filtered object and the full dir object, it won't try to read from the full dir object, right?

Reply:
CC @pmrowla

Reply:
> So, if we pass the content of filtered object and the full dir object, it won't try to read from the full dir object, right?

`shallow=bool(prefix)` looks correct to me.

If `shallow=True`, `status.compare_status` will not load a full dir/tree object, and assumes that you are explicitly passing every (raw) obj ID that you want it to query.

So if `obj_ids` here contains all the filtered child obj IDs, we want `shallow=True`.

For the unfiltered case (where `prefix` is unset), we do want it to load the tree (since `obj_ids` will only contain the top level tree obj ID, and has not been updated to contain any of the child obj IDs)